### PR TITLE
MEX -156

### PR DESCRIPTION
**Pull request by c4role**
## Describe your changes
When create or import a new world, I should be selected as currentWorld.
When you delete a world, the first world of the list should be selected as currentWorld.

## Checklist tests
You can create new world : check if it's the one selected and display.
Same test when import world.
Delete a world, check if selected world is the first of the list.

## Issue ticket number and link
[MEX-154] https://entsupport.gdapublic.fr/browse/MEX-154
[MEX-156] https://entsupport.gdapublic.fr/browse/MEX-156

These two bug tickets actually have the same resolution, so I handled them in the same ticket.

## Checklist before requesting a review (magic string, indentation, comment/documentation...)

- [x] I have detailed the tests to do in my feature/fix in order to prevent consequents regressions (must specify in **Checklist tests**)
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (API Doc etc...) - (must specify in **Description** for target version)
- [ ] If it is a consequent feature, I have added thorough tests.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to this project (must specify in **Description**)

### CODE REVIEW

File: src/main/resources/public/ts/controllers/minetest.controller.ts [hunk 1]
```diff
@@ -14,7 +14,7 @@ interface IViewModel {
     getLink(): string;
     getWiki(): string;
     getDownload(): string;
-    refreshWorldList(): Promise<void>;
+    refreshWorldList(world?): any;
```
Comment (BastienBenard) [inline src/main/resources/public/ts/controllers/minetest.controller.ts]: typage + est ce que world est obligatoire ?

File: src/main/resources/public/ts/controllers/minetest.controller.ts [hunk 2]
```diff
@@ -63,18 +60,19 @@ class Controller implements ng.IController, IViewModel {
         safeApply(this.$scope);
     }
 
-    async getWorld(): Promise<void> {
+    async getWorld(world?): Promise<void> {
```
Comment (BastienBenard) [inline src/main/resources/public/ts/controllers/minetest.controller.ts]: typage

File: src/main/resources/public/ts/controllers/minetest.controller.ts [hunk 3]
```diff
@@ -63,18 +60,19 @@ class Controller implements ng.IController, IViewModel {
...
         this.worlds.all = await minetestService.get(this.user_id, this.user_name);
-        this.setCurrentWorld();
+        if (world) {
+            let currentWorld = this.worlds.all.find(w => w._id === world._id)
```
Comment (BastienBenard) [inline src/main/resources/public/ts/controllers/minetest.controller.ts]: typage currentWorld et w

File: src/main/resources/public/ts/controllers/minetest.controller.ts [hunk 4]
```diff
@@ -63,18 +60,19 @@ class Controller implements ng.IController, IViewModel {
...
     }
 
-    setCurrentWorld(): void {
-        if (this.currentWorld) {
-            let worldId = this.currentWorld._id
-            this.currentWorld = this.worlds.all.filter(w => w._id === worldId)[0]
-        } else {
-            this.currentWorld = this.worlds.all[0]
-        }
+    setCurrentWorld(world?): void {
```
Comment (BastienBenard) [inline src/main/resources/public/ts/controllers/minetest.controller.ts]: il faut changé l'interface + typage

File: src/main/resources/public/ts/controllers/minetest.controller.ts [hunk 5]
```diff
@@ -102,9 +100,11 @@ class Controller implements ng.IController, IViewModel {
         return window.minetestDownload;
     }
 
-    async refreshWorldList(): Promise<void> {
-        await this.getWorld();
-        safeApply(this.$scope);
+    refreshWorldList(): (world) => any {
```
Comment (BastienBenard) [inline src/main/resources/public/ts/controllers/minetest.controller.ts]: typage de world. Ce n'est pas possible de changé le any ?

File: src/test/java/fr/openent/minetest/services/DefaultWorldServiceTest.java [hunk 6]
```diff
@@ -70,7 +70,7 @@ public void testCreateImportWorld(TestContext context) {
         try {
             Whitebox.invokeMethod(worldService, "importWorld", world, userInfos);
         } catch (Exception e) {
-            context.assertNull(e);
+            context.assertNotNull(e);
```
Comment (BastienBenard) [inline src/test/java/fr/openent/minetest/services/DefaultWorldServiceTest.java]: Il faut que tu remplace le expectedWorld pour qu'il passe.

**Review by BastienBenard — COMMENTED** [review]
[fix(controller): #MX-154 #MX-156 Update currentWord](https://github.com/OPEN-ENT-NG/minetest/pull/35/commits/a930fc704701087788e595c9c20afb08b52c620e) c'est un commit pour 2 tickets ?

File: src/main/resources/public/ts/controllers/minetest.controller.ts (hunk 1 above)
Comment (c4role) [reply src/main/resources/public/ts/controllers/minetest.controller.ts]: Non, world est optionnel

File: src/main/resources/public/ts/controllers/minetest.controller.ts (hunk 5 above)
Comment (c4role) [reply src/main/resources/public/ts/controllers/minetest.controller.ts]: Non c'est pas possible

File: src/main/resources/public/ts/directives/create-world/create-world.directive.ts [hunk 7]
```diff
@@ -77,15 +82,18 @@ class Controller implements ng.IController, IViewModel {
             title: this.world.title,
             address: this.formatAddress()
         }
-        minetestService.create(this.world)
-            .then(() => {
+        minetestService.create(this.newWorld)
+            .then((world: AxiosResponse) => {
                 toasts.confirm('minetest.world.create.confirm');
+                if (this.$scope.$parent.$eval(this.$scope['vm']['onCreateWorld'])(world.data))
+                    this.$scope.$parent.$eval(this.$scope['vm']['onCreateWorld'])(world.data);
```
Comment (qpic-cgi) [inline src/main/resources/public/ts/directives/create-world/create-world.directive.ts:89]: je ne pense pas que cette ligne soit utile, vu que tu le check dans le if, si ?

File: src/main/resources/public/ts/directives/delete-world/delete-world.directive.ts [hunk 8]
```diff
@@ -46,18 +46,21 @@ class Controller implements ng.IController, IViewModel {
...
     }
 
     async deleteImportWorld(): Promise<void> {
         minetestService.deleteImportWorld(this.world)
             .then(() => {
                 toasts.confirm('minetest.world.delete.confirm');
                 this.closeDeleteLightbox();
-                this.$scope.$eval(this.$scope['vm']['onDeleteWorld']());
+                if (this.$scope.$parent.$eval(this.$scope['vm']['onDeleteWorld'])(this.world))
+                    this.$scope.$parent.$eval(this.$scope['vm']['onDeleteWorld'])(this.world);
```
Comment (qpic-cgi) [inline src/main/resources/public/ts/directives/delete-world/delete-world.directive.ts:63]: je ne pense pas que cette ligne soit utile, vu que tu le check dans le if, si ?

**Comment by c4role** [comment]
> [fix(controller): #MX-154 #MX-156 Update currentWord](https://github.com/OPEN-ENT-NG/minetest/pull/35/commits/a930fc704701087788e595c9c20afb08b52c620e) c'est un commit pour 2 tickets ?

Oui j'ai expliqué pourquoi dans la description de ma PR

File: src/main/resources/public/ts/directives/import-world/import-world.directive.ts [hunk 9]
```diff
@@ -84,11 +89,12 @@ class Controller implements ng.IController, IViewModel {
             port: parseInt(String(this.world.port)),
             isExternal: true
         }
-        minetestService.import(this.world)
-            .then(() => {
+        minetestService.import(this.newImportWorld)
+            .then((world: AxiosResponse) => {
                 toasts.confirm('minetest.world.import.confirm');
+                if (this.$scope.$parent.$eval(this.$scope['vm']['onImportWorld'])(world.data))
+                    this.$scope.$parent.$eval(this.$scope['vm']['onImportWorld'])(world.data);
```
Comment (qpic-cgi) [inline src/main/resources/public/ts/directives/import-world/import-world.directive.ts:96]: je ne pense pas que cette ligne soit utile, vu que tu le check dans le if, si ?

File: src/main/resources/public/ts/controllers/minetest.controller.ts (hunk 5 above)
Comment (BastienBenard) [reply src/main/resources/public/ts/controllers/minetest.controller.ts]: https://github.com/OPEN-ENT-NG/minetest/pull/35#discussion_r928831463 du coups world est optionnel

Comment (c4role) [reply src/main/resources/public/ts/controllers/minetest.controller.ts]: En effet, merci :)

**Pull request merged**